Riot/Modules/MatrixKit/Utils/EventFormatter/MarkdownToHTMLRenderer.swift

```diff
@@ -50,3 +57,103 @@ public class MarkdownToHTMLRendererHardBreaks: MarkdownToHTMLRenderer {
+        case CMARK_NODE_EMPH:
+            return "_"
+        case CMARK_NODE_STRONG:
+            return "__"
+        default:
+            return ""
+        }
+    }
+
+    /// Repairs links that were broken down by markdown formatting.
+    /// Should be used on the first node of libcmark's AST
+    /// (e.g. the object returned by DownASTRenderer.stringToAST).
+    func repairLinks() {
+        let iterator = cmark_iter_new(self)
+        var text = ""
+        var isInParagraph = false
+        var previousNode: CMarkNode?
+        var shouldUnlinkFormattingMode = false
+        var event: cmark_event_type?
+        while event != CMARK_EVENT_DONE {
+            event = cmark_iter_next(iterator)
+
+            guard let node = cmark_iter_get_node(iterator) else { return }
+
+            if node.type == CMARK_NODE_PARAGRAPH {
+                if event == CMARK_EVENT_ENTER {
+                    isInParagraph = true
+                } else {
+                    isInParagraph = false
+                    text = ""
+                }
+            }
+
+            if isInParagraph {
+                switch node.type {
+                case CMARK_NODE_SOFTBREAK,
+                     CMARK_NODE_LINEBREAK:
+                    text = ""
+                case CMARK_NODE_TEXT:
+                    if let literal = node.literal {
+                        text += literal
+                        // Reset text if it ends up with a whitespace.
+                        if text.last?.isWhitespace == true {
+                            text = ""
+                        }
+                        // Only the last part could be a link conflicting with next node.
+                        text = String(text.split(separator: " ").last ?? "")
+                    }
+                case CMARK_NODE_EMPH where previousNode?.type == CMARK_NODE_TEXT,
+                     CMARK_NODE_STRONG where previousNode?.type == CMARK_NODE_TEXT:
+                    if event == CMARK_EVENT_ENTER {
+                        if !text.containedUrls.isEmpty,
+                           let childLiteral = node.pointee.first_child.literal {
+                            // If current text is a link, the formatted text is reverted back to a
+                            // plain text as a part of the link.
+                            let symbol = node.formattingSymbol
+                            let nonFormattedText = "\(symbol)\(childLiteral)\(symbol)"
+                            let replacementTextNode = cmark_node_new(CMARK_NODE_TEXT)
+                            cmark_node_set_literal(replacementTextNode, nonFormattedText)
+                            cmark_node_insert_after(previousNode, replacementTextNode)
+                            cmark_node_set_literal(node.pointee.first_child, "")
+                            let newIterator = cmark_iter_new(node)
+                            _ = cmark_iter_next(newIterator)
+                            cmark_node_unlink(node)
```

Comment:
these actually don't free mem (at least not immediately), I will update for that, 

I'll take a look if I can as well make sure that created node that I link manually to the AST are properly freed (hopefully they are, cause freeing manually afterwards would look terrible)